Html email templates

Neriderc opened the pull request:
Added HTML email templates for notification email and password reset email. It lists the name of the post with a link, and a link to the feed at the bottom. I'm no UI expert but I think it looks ok. E.g.:

![screen shot of email from zusam - list of names of new posts + a view feed button](https://github.com/user-attachments/assets/99590baf-6666-4d68-a91c-5512860c0451)

I believe this resolves #48, although that user is not using English and other templates would need to be created for other languages. These templates are only for English in this pull request. 

Text emails are not affected. If a user has an email client where they have requested text emails, they will be the same as before.
 
This PR also includes an earlier database write to mitigate duplicate emails when the process is running multiple times at the same time.

Neriderc commented:
I'm keen to understand a bit more about when the notification email command is triggered as it confused my a bit. I believe this PR is working properly, but emails seem to be triggered immediately rather than when the cron is run.

Does some process trigger sending emails or running cron when a notification is created? Given that previously emails were only sent once per hour at the most, it seems odd that posting a new post causes the email command to trigger.

nrobinaubertin commented:
> I'm keen to understand a bit more about when the notification email command is triggered as it confused me a bit. I believe this PR is working properly, but emails seem to be triggered immediately rather than when the cron is run.
> 
> Does some process trigger sending emails or running cron when a notification is created? Given that previously emails were only sent once per hour at the most, it seems odd that posting a new post causes the email command to trigger.

The cron command was made to mimic the idea behind wordpress's task runner. If zusam is installed on a system without cron (like a web share), we need a way to run it regularly. The cron command is running every time the API finishes to respond to a call, via the event listener [here](https://github.com/zusam/zusam/blob/master/api/src/EventListener/TerminateListener.php).

But you also can additionally add the command to your system cron to be sure that it executes even if no endpoint is called. 

When launched, the cron command gets some information about the possible tasks [here](https://github.com/zusam/zusam/blob/master/api/src/Command/Cron.php#L39).  
"light" tasks can be executed all the time and "heavy" ones are reserved for some (low traffic) hours defined [here](https://github.com/zusam/zusam/blob/master/api/.env#L44).  

It then loops over the possible tasks and executes the first one that has a period under `now-lastExecutionDateTime` (code [here](https://github.com/zusam/zusam/blob/master/api/src/Command/Cron.php#L180)).  
The command execution data is stored in a table named `system` (see [here](https://github.com/zusam/zusam/blob/master/api/src/Entity/System.php)).

Now that I'm writing all this, I'm realizing that having put the period for the notification emails to 0 makes it execute every time the cron command is called (and nothing else is going on). So we need to make sure that the conditions defined [here](https://github.com/zusam/zusam/blob/master/api/src/Command/NotificationEmails.php#L54) are safeguarding against sending them too often.  
Having the period set to 0 also means that it's going to steal all the executions from commands down the line so the [task definition](https://github.com/zusam/zusam/blob/master/api/src/Command/Cron.php#L51) should be put at the end to have the lowest priority

Neriderc commented:
Ah right. So what I did was ->flush() to update the database after the last check time was updated so that it wouldn't send duplicate emails (waiting until the end of the run to update the database could cause two or three emails to run, which I'm assuming is related to this cron setup causing multiple to run at the same time - I confirmed each run was only sending one email). But updating the database straight after setting the new check time, I have not seen any duplicates anymore.

Is your suggestion to (in addition to lowering the priority) exit if the cron has run in the last X amount of time? Rather than a change in the NofiticationEmail.php criteria, would it make sense to change this:
```
[
                'name' => 'zusam:notification:emails',
                'period' => intval($this->params->get('cron.notification.emails')),
                'type' => 'light',
                'options' => [
                    '--log-send' => true,
                ],
            ],
```

To have a minimum period of 1? Or even leave this and instead set `cron.notification.emails` (that we have renamed) to default to 1 instead of 0?

It seems wrong to have the logic for deciding if a task is running too often split across two different places (Cron.php and NotificationEmails.php).

nrobinaubertin commented:
Maybe we should go back to the previous logic for `hourly` or slower notifications and send a notification email directly from the controller when a new message is created only for the users that have set notifications to `immediately` ?  
I feel like this would be the best of both worlds without too much added complexity

nrobinaubertin commented:
Another thing: I have nothing against the proposed design of the html emails but I know that html emails can be really tricky to get right for all types of client. The last (and unique) time I had to do that in the past, I used [mjml](https://mjml.io/).  

I'm willing to accept your PR as a first step in having html emails but I feel like we should use a robust solution in the future.

Neriderc commented:
> Maybe we should go back to the previous logic for `hourly` or slower notifications and send a notification email directly from the controller when a new message is created only for the users that have set notifications to `immediately` ? I feel like this would be the best of both worlds without too much added complexity

I don't think I follow how this would work. The previous logic would only allow the task to run once per hour. In this case, how do we get the email notifications to send for those that have chosen Immediately? The per-user checks happen in the NotificationEmail command while this command is not called at all if we only allow it to run once per hour using the previous logic.

I think the current setup (skipping the user if it's not the appropriate time for their notifications) is about the same thing. But I'd be happy to hear more about what setup you were envisaging.

> Another thing: I have nothing against the proposed design of the html emails but I know that html emails can be really tricky to get right for all types of client. The last (and unique) time I had to do that in the past, I used [mjml](https://mjml.io/).
> 
> I'm willing to accept your PR as a first step in having html emails but I feel like we should use a robust solution in the future.

These emails are not too complicated and I have tested looking at them on mobile as well as desktop, and they look ok for my specific devices. But I am happy to have a play with mjml for a future PR, I wasn't aware it existed until now. Do you think #48 should be kept open for this, or closed and a new issue created?

nrobinaubertin commented:
> > Maybe we should go back to the previous logic for `hourly` or slower notifications and send a notification email directly from the controller when a new message is created only for the users that have set notifications to `immediately` ? I feel like this would be the best of both worlds without too much added complexity
> 
> I don't think I follow how this would work. The previous logic would only allow the task to run once per hour. In this case, how do we get the email notifications to send for those that have chosen Immediately? The per-user checks happen in the NotificationEmail command while this command is not called at all if we only allow it to run once per hour using the previous logic.
> 
> I think the current setup (skipping the user if it's not the appropriate time for their notifications) is about the same thing. But I'd be happy to hear more about what setup you were envisaging.

My idea was that we could have two paths for sending emails. Nothing restricts us to have two commands for emails (or another way to decide when to send emails).
But this would complicate things, I agree.

What about adding a new task type: `always`. Those would be executed every time, *after* that the CronCommand has tried/executed another task type.  
For example, cronCommand would wake up, see that it needs to execute the image conversion, run it and then run all the `always` tasks one by one.
Another example: cronCommands wakes up, sees that there is already a task running but executes all the `always` tasks one by one.  
Those tasks should succeed/fail fast, be light on resources and would need to be checked *very* often. Notifications is a perfect candidate

> Do you think #48 should be kept open for this, or closed and a new issue created?

I think that we can delegate using a email templating framework to another issue. IMO: better start somewhere like you did and improve later.

Neriderc commented:
Ok I've created #100 to track using an email templating framework.

> What about adding a new task type: always

This is a good idea. Even if some videos are busy being converted, which might take 30 mins or more depending on the hardware and the video, we still want the notifications to send. 

Is the key part here the `defined('TASK_RUNNING') ` that checks [here](https://github.com/zusam/zusam/blob/master/api/src/Command/Cron.php#L129) if a task is running? Allow it to continue with the always checks in this case?

nrobinaubertin commented:
> Is the key part here the `defined('TASK_RUNNING') ` that checks [here](https://github.com/zusam/zusam/blob/master/api/src/Command/Cron.php#L129) if a task is running? Allow it to continue with the always checks in this case?

Not quite: the `defined` construct is use here to avoid running another task by calling cron again in the same process. You can see that I set it [a few lines after](https://github.com/zusam/zusam/blob/master/api/src/Command/Cron.php#L135).  
I think that the key part is [here](https://github.com/zusam/zusam/blob/master/api/src/Command/Cron.php#L143). Instead of doing nothing (returning false), we should call the "execute `always` tasks" function. This function should also be called before the `return true` of the cron.

Neriderc commented:
Thanks, I'll have a play when I get a chance and see if I can work it out.

Neriderc commented:
I've added a commit that I think does what you were explaining.

- Set notification cron type to "always".
- If we would otherwise quit because of a long running task, instead set a variable $onlyRunAlwaysTasks and continue.
- Don't update the last task - we don't want the following cron to start repeating itself so this doesn't count as running a task.
- We only run "always" tasks when we get to the task running part

This makes sense in my mind, hopefully it's what you had in mind. I have tested and I don't get duplicate emails, and I loaded up some image/video conversion stuff and it seems to send the notifications now if I make a new post while that is running.

If you'd prefer I split out the calling of the always tasks into a new function I could do that instead.

nrobinaubertin commented on api/src/Command/Cron.php:
```diff
@@ -140,31 +140,43 @@ public function runTask(): bool
+        if (!$onlyRunAlwaysTasks) {
+            // if a task is running but it's old, it's probably a lockup.
+            // continue but log it
+            if (!$onlyRunAlwaysTasks && isset($last_task_timestamp) && $last_task_timestamp < time() - intval($this->params->get('max_task_lock_duration'))) {
```
We don't need the repeated `!$onlyRunAlwaysTasks` here

nrobinaubertin commented on api/src/Command/Cron.php:
```diff
@@ -140,31 +140,43 @@ public function runTask(): bool
+
+            // If we are only doing "always" tasks, skip others
+            if ($onlyRunAlwaysTasks
+                && isset($task['type'])
```
Let's move the `isset` before and make it so that if the type is not defined, we don't run it

Neriderc replied:
I think I have updated this to what you're meaning.

nrobinaubertin reviewed (commented):
Minor comments. I feel like this is enough for this PR but I will probably revisit the cron logic in a future PR and tag you as reviewer. For example: I think that it makes sense to stop the cron process after one successful task (to avoid long running processes).

Neriderc commented:
I've added a new commit to address your feedback. I've also changed where the $idle_hours default is declared as this was picked up in the phpstan check after that section was modified.

nrobinaubertin commented on api/src/Command/Cron.php:
```diff
@@ -140,35 +142,43 @@ public function runTask(): bool
+
+            // If we are only doing "always" tasks, skip others
+            if (!isset($task['type'])
+                || $onlyRunAlwaysTasks && 'always' !== $task['type']) {
```
I just learned that `&&` has higher precedence than `||` in php so your condition works. But it's maybe more readable if you add a pair of parenthesis around `$onlyRunAlwaysTasks && 'always' !== $task['type']` ?

Neriderc replied:
Haha I was going to do that but I knew `&&` has higher precedence in PHP (I think this was true for most languages?) so I thought you might consider it unnecessary.

I've used the style from something further down so it should be consistent now.

nrobinaubertin reviewed (approved):
Just a nitpick

nrobinaubertin commented:
Thank you for the HTML templates !
